satpy/cf/datasets.py

```diff
@@ -0,0 +1,281 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2017-2023 Satpy developers
+#
+# This file is part of satpy.
+#
+# satpy is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# satpy is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Utility to generate a CF-compliant Datasets."""
+import logging
+import warnings
+from collections import defaultdict
+
+import xarray as xr
+
+from satpy.writers.cf_writer import CF_DTYPES, CF_VERSION
+
+logger = logging.getLogger(__name__)
+
+
+def _get_extra_ds(dataarray, keys=None):
+    """Get the ancillary_variables DataArrays associated to a dataset."""
+    dict_datarrays = {}
+    # Retrieve ancillary variable datarrays
+    for ancillary_dataarray in dataarray.attrs.get("ancillary_variables", []):
+        ancillary_variable = ancillary_dataarray.name
+        if keys and ancillary_variable not in keys:
+            keys.append(ancillary_variable)
+            dict_datarrays.update(_get_extra_ds(ancillary_dataarray, keys=keys))
+    # Add input dataarray
+    dict_datarrays[dataarray.attrs["name"]] = dataarray
+    return dict_datarrays
+
+
+def _get_groups(groups, list_datarrays):
+    """Return a dictionary with the list of xr.DataArray associated to each group.
+
+    If no groups (groups=None), return all DataArray attached to a single None key.
+    Else, collect the DataArrays associated to each group.
+    """
+    if groups is None:
+        grouped_dataarrays = {None: list_datarrays}
+    else:
+        grouped_dataarrays = defaultdict(list)
+        for datarray in list_datarrays:
+            for group_name, group_members in groups.items():
+                if datarray.attrs["name"] in group_members:
+                    grouped_dataarrays[group_name].append(datarray)
+                    break
+    return grouped_dataarrays
```

Comment:
```suggestion
    if groups is None:
        return {None: list_datarrays}

    grouped_dataarrays = defaultdict(list)
    for datarray in list_datarrays:
        for group_name, group_members in groups.items():
            if datarray.attrs["name"] in group_members:
                grouped_dataarrays[group_name].append(datarray)
                break
    return grouped_dataarrays
```

I feel like an additional generator function for the for loops would clean this up a bit too, but not as important.

Reply:
I implemented also the generator ;)